File: py/abstraction/facade/jobs.py
"""Validated durable job binding. Transport cancellation changes waiting only."""
import threading
import time
from abstraction.ipc import FrameTransport, Library
from abstraction.job.acceptance import rec as wire

class JobError(RuntimeError):
    def __init__(self, code, message):
        super().__init__(message)
        self.code = code

class ResultCopyError(RuntimeError):
    """confirmed bytes were written; cause retains the original typed failure."""
    def __init__(self, confirmed, cause):
        super().__init__(str(cause))
        self.confirmed, self.cause = confirmed, cause
        self.code = getattr(cause, "code", None)

def require(condition, code, message):
    if not condition:
        raise JobError(code, message)

def text(value):
    if not isinstance(value, str) or not value:
        return False
    try:
        value.encode("utf-8")
        return True
    except UnicodeError:
        return False

def guarantees(values, code="invalid_submission"):
    values = tuple(values)
    require(all(text(g) for g in values) and len(set(values)) == len(values),
            code, "empty or duplicate guarantee")
    return values

def identity(value):
    require(isinstance(value, wire.RequestIdentity) and text(value.key) and text(value.history_epoch),
            "invalid_submission", "explicit key and history epoch required")

class Owner:
    def __init__(self, value):
        self.value, self.lock = value, threading.Lock()
    def bind(self, value):
        with self.lock:
            require(text(value) and (not self.value or self.value == value),
                    "invalid_acceptance", "logical owner absent or changed")
            self.value = value

class Jobs(wire.RecoverableAcceptance, wire.OperationControl):
    def __init__(self, transport, *, required_guarantees=(), expected_owner="", _owner=None):
        require(isinstance(expected_owner, str), "invalid_submission", "invalid expected owner")
        self._transport = transport
        self._required = guarantees(required_guarantees)
        self._owner = _owner if _owner is not None else Owner(expected_owner)
        self._acceptance = wire.RecoverableAcceptanceClient(transport)
        self._operations = wire.OperationControlClient(transport)

    @classmethod
    def restore(cls, endpoint, owner, *, required_guarantees=(), library=None, **waiting):
        """Use caller-retained endpoint/owner/complete requirements after restart."""
        require(text(owner), "invalid_submission", "persisted logical owner required")
        library = library if library is not None else (waiting["cancellation"]._library if waiting.get("cancellation") is not None else Library())
        return cls(FrameTransport(library, endpoint, max_frame=2*1024*1024, **waiting),
                   required_guarantees=required_guarantees, expected_owner=owner)

    @classmethod
    def restore_installed(cls, endpoint, owner, *, required_guarantees=(), library=None,
                          timeout=5.0, deadline=None, cancellation=None):
        """Restore retained work using independently selected installation trust."""
        require(text(owner), "invalid_submission", "persisted logical owner required")
        library = library if library is not None else (cancellation._library if cancellation is not None else Library())
        # Validate waiting arguments before computing a composite selection budget.
        FrameTransport(library, endpoint, timeout=timeout, deadline=deadline, cancellation=cancellation)
        end = deadline if deadline is not None else time.monotonic() + timeout
        server = library.select_runtime(timeout=timeout, deadline=end, cancellation=cancellation)
        return cls.restore(endpoint, owner, required_guarantees=required_guarantees, library=library,
                           server=server, timeout=timeout, deadline=deadline, cancellation=cancellation)

    @property
    def endpoint(self):
        return self._transport.endpoint.decode("utf-8")

    @property
    def owner(self):
        with self._owner.lock:
            return self._owner.value

    def with_waiting(self, *, timeout=None, deadline=None, cancellation=None):
        """Fresh waiting policy, same fixed endpoint and synchronized owner."""
        t = self._transport
        transport = FrameTransport(t.library, self.endpoint,
                                   timeout=t.timeout if timeout is None else timeout,
                                   deadline=deadline, cancellation=cancellation, max_frame=t.max_frame, server=t.server)
        return type(self)(transport, required_guarantees=self._required, _owner=self._owner)

    def GetHistoryWindow(self):
        result = self._acceptance.GetHistoryWindow()
        require(text(result.history_epoch) and result.minimum_retention_ms > 0,
                "invalid_acceptance", "invalid history window")
        self._owner.bind(result.logical_owner)
        return result

    def _receipt(self, receipt, requested, required, *, pin=True):
        identity(requested)
        require(receipt is not None and receipt.identity.key == requested.key
                and receipt.identity.history_epoch == requested.history_epoch
                and text(receipt.operation_id) and text(receipt.logical_owner)
                and receipt.history_retention_ms > 0,
                "invalid_acceptance", "missing or mismatched receipt")
        accepted = guarantees(receipt.accepted_guarantees, "invalid_acceptance")
        require(set(required).issubset(accepted), "invalid_acceptance", "receipt weakens requirements")
        if pin:
            self._owner.bind(receipt.logical_owner)

    def _result(self, result, requested, required):
        if result.outcome == "accepted":
            self._receipt(result.receipt, requested, required)
        else:
            require(result.outcome in ("definitely_not_accepted", "unknown", "key_conflict", "forbidden", "invalid")
                    and result.receipt is None, "invalid_acceptance", "inconsistent acceptance response")
        return result

    def Submit(self, submission):
        identity(submission.identity)
        require(text(submission.kind), "invalid_submission", "kind required")
        requested = list(guarantees(submission.required_guarantees))
        requested.extend(g for g in self._required if g not in requested)
        outgoing = wire.Submission(identity=submission.identity, kind=submission.kind,
                                   spec=submission.spec, required_guarantees=requested)
        return self._result(self._acceptance.Submit(outgoing), submission.identity, requested)

    def Reconcile(self, requested):
        identity(requested)
        return self._result(self._acceptance.Reconcile(requested), requested, self._required)

    def CancelWork(self, requested):
        identity(requested)
        result = self._acceptance.CancelWork(requested)
        require(result.outcome in wire.CANCELLATIONOUTCOME_NAMES,
                "invalid_cancellation", "unknown cancellation outcome")
        return result

    def _snapshot(self, snapshot, requested, required, *, pin=True):
        require(snapshot is not None and snapshot.state in
                ("pending", "running", "transferred", "complete", "failed", "cancelled")
                and snapshot.progress.done >= 0 and snapshot.progress.total >= 0,
                "invalid_observation", "inconsistent progress/state")
        require(snapshot.failure is None or snapshot.failure.classification in
                ("retryable", "permanent", "unknown"), "invalid_observation", "invalid last failure")
        self._receipt(snapshot.receipt, requested, required, pin=pin)

    def ObserveWork(self, requested):
        identity(requested)
        result = self._operations.ObserveWork(requested)
        if result.outcome == "observed":
            self._snapshot(result.snapshot, requested, self._required)
        else:
            require(result.outcome in ("unknown", "forbidden", "invalid", "definitely_not_accepted")
                    and result.snapshot is None, "invalid_observation", "inconsistent observation")
        return result

    def ReadResult(self, requested, offset, max_bytes):
        identity(requested)
        require(type(offset) is int and 0 <= offset < 2**63 and type(max_bytes) is int
                and 1 <= max_bytes <= 65536, "invalid_request", "invalid result range")
        result = self._operations.ReadResult(requested, offset, max_bytes)
        if result.outcome != "data":
            require(result.outcome in ("not_ready", "unavailable", "unsupported", "unknown", "forbidden", "invalid")
                    and result.chunk is None, "invalid_result", "inconsistent result refusal")
            return result
        c = result.chunk
        require(c is not None and c.offset == offset and c.total >= offset,
                "invalid_result", "invalid chunk offset or total")
        n = len(c.data)
        require(n <= max_bytes and n <= c.total-offset and c.eof == (n == c.total-offset)
                and (n > 0 or c.eof), "invalid_result", "invalid chunk bounds or EOF")
        self._receipt(c.receipt, requested, self._required)
        return result

    def CopyResult(self, requested, destination):
        """Copy under one call_scope() budget; preserve confirmed partial bytes.

        Injected transports must provide call_scope() with a shared deadline
        across its exchanges. The reusable binding is never mutated.
        """
        written, operation, total = 0, None, None
        try:
            scope = getattr(self._transport, "call_scope", None)
            require(callable(scope), "unsupported_waiting", "CopyResult transport requires call_scope")
            scoped = Jobs(scope(), required_guarantees=self._required,
                          _owner=self._owner)
            while True:
                result = scoped.ReadResult(requested, written, 65536)
                require(result.outcome == "data", result.outcome, "result copy unavailable")
                c = result.chunk
                if operation is None:
                    operation, total = c.receipt.operation_id, c.total
                require(c.receipt.operation_id == operation and c.total == total,
                        "invalid_result", "result identity or total changed")
                if c.data:
                    n = destination.write(c.data)
                    require(type(n) is int and 0 <= n <= len(c.data), "invalid_writer", "invalid write count")
                    written += n
                    require(n == len(c.data), "short_write", "result writer returned a short write")
                if c.eof:
                    return written
        except Exception as error:
            raise ResultCopyError(written, error) from error

class Inventory(wire.JobInventory):
    def __init__(self, transport, *, required_guarantees=(), expected_owner="", _owner=None):
        self._binding = Jobs(transport, required_guarantees=required_guarantees,
                             expected_owner=expected_owner, _owner=_owner)
        self._transport = transport

    @property
    def endpoint(self):
        return self._binding.endpoint

    @property
    def owner(self):
        return self._binding.owner

    def with_waiting(self, **waiting):
        binding = self._binding.with_waiting(**waiting)
        return Inventory(binding._transport, required_guarantees=binding._required, _owner=binding._owner)

    def ListWork(self, cursor, limit):
        require(isinstance(cursor, str) and len(cursor.encode("utf-8")) <= 128
                and type(limit) is int and 1 <= limit <= 64,
                "invalid_request", "invalid inventory cursor or limit")
        page = wire.JobInventoryClient(self._transport).ListWork(cursor, limit)
        if page.outcome != "page":
            require(page.outcome in ("gap", "forbidden", "invalid", "unavailable")
                    and not page.snapshots and not page.next and not page.complete,
                    "invalid_inventory", "inconsistent inventory refusal")
            return page
        require(len(page.snapshots) <= limit and len(page.next.encode("utf-8")) <= 128
                and ((page.complete and not page.next) or
                     (not page.complete and page.next and page.next != cursor)),
                "invalid_inventory", "inconsistent inventory continuation")
        seen, owner = set(), None
        for snapshot in page.snapshots:
            # Inventory negotiation cannot impose new guarantees on older work.
            self._binding._snapshot(snapshot, snapshot.receipt.identity, (), pin=False)
            receipt = snapshot.receipt
            require(receipt.operation_id not in seen and (owner is None or owner == receipt.logical_owner),
                    "invalid_inventory", "duplicate operation or mixed owners")
            seen.add(receipt.operation_id)
            owner = receipt.logical_owner
        if owner is not None:
            self._binding._owner.bind(owner)
        return page

File: rust-jobs/src/jobs.rs
//! Fixed provider binding. Transport errors leave acceptance unknown; no retry is issued.
pub use abstraction_job_api as wire;
use std::{
    io::Write,
    sync::{Arc, Mutex},
};
use wire::{JobInventory, OperationControl, RecoverableAcceptance};
#[derive(Debug)]
pub enum Error<E> {
    Call(wire::CallError<E>),
    Transport(E),
    Invalid(&'static str),
    Outcome(String),
    Writer(std::io::Error),
}
#[derive(Debug)]
pub struct CopyError<E> {
    pub confirmed: u64,
    pub cause: Error<E>,
}
fn require<E>(ok: bool, reason: &'static str) -> Result<(), Error<E>> {
    if ok {
        Ok(())
    } else {
        Err(Error::Invalid(reason))
    }
}
fn id(v: &wire::RequestIdentity) -> wire::RequestIdentity {
    wire::RequestIdentity {
        key: v.key.clone(),
        history_epoch: v.history_epoch.clone(),
    }
}
fn identity<E>(v: &wire::RequestIdentity) -> Result<(), Error<E>> {
    require(
        !v.key.is_empty() && !v.history_epoch.is_empty(),
        "request identity",
    )
}
#[derive(Clone)]
pub struct Jobs<T> {
    transport: T,
    owner: Arc<Mutex<String>>,
    required: Vec<String>,
}
impl<T: wire::FrameTransport + Clone> Jobs<T> {
    pub fn new(
        transport: T,
        expected_owner: String,
        required: Vec<String>,
    ) -> Result<Self, Error<T::Error>> {
        require(super::distinct(&required), "required guarantees")?;
        Ok(Self {
            transport,
            owner: Arc::new(Mutex::new(expected_owner)),
            required,
        })
    }
    pub fn owner(&self) -> String {
        self.owner.lock().unwrap().clone()
    }
    fn pin(&self, owner: &str) -> Result<(), Error<T::Error>> {
        let mut current = self.owner.lock().unwrap();
        require(
            !owner.is_empty() && (current.is_empty() || *current == owner),
            "logical owner",
        )?;
        *current = owner.into();
        Ok(())
    }
    fn receipt(
        &self,
        r: &wire::Receipt,
        wanted: &wire::RequestIdentity,
        required: &[String],
    ) -> Result<(), Error<T::Error>> {
        identity(wanted)?;
        require(
            r.identity.key == wanted.key
                && r.identity.history_epoch == wanted.history_epoch
                && !r.operation_id.is_empty()
                && !r.logical_owner.is_empty()
                && r.history_retention_ms > 0
                && super::distinct(&r.accepted_guarantees)
                && required.iter().all(|g| r.accepted_guarantees.contains(g)),
            "receipt",
        )
    }
    fn acceptance(
        &self,
        r: wire::AcceptanceResult,
        wanted: &wire::RequestIdentity,
        required: &[String],
    ) -> Result<wire::AcceptanceResult, Error<T::Error>> {
        if r.outcome == "accepted" {
            let receipt = r
                .receipt
                .as_ref()
                .ok_or(Error::Invalid("missing receipt"))?;
            self.receipt(receipt, wanted, required)?;
            self.pin(&receipt.logical_owner)?;
        } else {
            require(
                matches!(
                    r.outcome.as_str(),
                    "definitely_not_accepted"
                        | "unknown"
                        | "key_conflict"
                        | "forbidden"
                        | "invalid"
                ) && r.receipt.is_none(),
                "acceptance outcome",
            )?;
        }
        Ok(r)
    }
    pub fn history_window(&self) -> Result<wire::HistoryWindow, Error<T::Error>> {
        let r = wire::RecoverableAcceptanceClient::new(self.transport.clone())
            .GetHistoryWindow()
            .map_err(Error::Call)?;
        require(
            !r.history_epoch.is_empty() && r.minimum_retention_ms > 0,
            "history window",
        )?;
        self.pin(&r.logical_owner)?;
        Ok(r)
    }
    pub fn submit(
        &self,
        mut submission: wire::Submission,
    ) -> Result<wire::AcceptanceResult, Error<T::Error>> {
        identity(&submission.identity)?;
        require(
            !submission.kind.is_empty() && super::distinct(&submission.required_guarantees),
            "submission",
        )?;
        for g in &self.required {
            if !submission.required_guarantees.contains(g) {
                submission.required_guarantees.push(g.clone());
            }
        }
        let wanted = id(&submission.identity);
        let required = submission.required_guarantees.clone();
        let r = wire::RecoverableAcceptanceClient::new(self.transport.clone())
            .Submit(submission)
            .map_err(Error::Call)?;
        self.acceptance(r, &wanted, &required)
    }
    pub fn reconcile(
        &self,
        wanted: &wire::RequestIdentity,
    ) -> Result<wire::AcceptanceResult, Error<T::Error>> {
        identity(wanted)?;
        let r = wire::RecoverableAcceptanceClient::new(self.transport.clone())
            .Reconcile(id(wanted))
            .map_err(Error::Call)?;
        self.acceptance(r, wanted, &self.required)
    }
    pub fn cancel_work(
        &self,
        wanted: &wire::RequestIdentity,
    ) -> Result<wire::CancellationResult, Error<T::Error>> {
        identity(wanted)?;
        wire::RecoverableAcceptanceClient::new(self.transport.clone())
            .CancelWork(id(wanted))
            .map_err(Error::Call)
    }
    fn snapshot(
        &self,
        s: &wire::OperationSnapshot,
        wanted: &wire::RequestIdentity,
        required: &[String],
    ) -> Result<(), Error<T::Error>> {
        require(
            matches!(
                s.state.as_str(),
                "pending" | "running" | "transferred" | "complete" | "failed" | "cancelled"
            ) && s.progress.done >= 0
                && s.progress.total >= 0
                && s.failure.as_ref().map_or(true, |f| {
                    matches!(
                        f.classification.as_str(),
                        "retryable" | "permanent" | "unknown"
                    )
                }),
            "observation",
        )?;
        self.receipt(&s.receipt, wanted, required)
    }
    pub fn observe(
        &self,
        wanted: &wire::RequestIdentity,
    ) -> Result<wire::ObservationResult, Error<T::Error>> {
        identity(wanted)?;
        let r = wire::OperationControlClient::new(self.transport.clone())
            .ObserveWork(id(wanted))
            .map_err(Error::Call)?;
        if r.outcome == "observed" {
            let s = r
                .snapshot
                .as_ref()
                .ok_or(Error::Invalid("missing snapshot"))?;
            self.snapshot(s, wanted, &self.required)?;
            self.pin(&s.receipt.logical_owner)?;
        } else {
            require(
                matches!(
                    r.outcome.as_str(),
                    "unknown" | "forbidden" | "invalid" | "definitely_not_accepted"
                ) && r.snapshot.is_none(),
                "observation outcome",
            )?;
        }
        Ok(r)
    }
    pub fn read_result(
        &self,
        wanted: &wire::RequestIdentity,
        offset: i64,
        max_bytes: i64,
    ) -> Result<wire::ResultRead, Error<T::Error>> {
        identity(wanted)?;
        require(
            offset >= 0 && (1..=65536).contains(&max_bytes),
            "result range",
        )?;
        let r = wire::OperationControlClient::new(self.transport.clone())
            .ReadResult(id(wanted), offset, max_bytes)
            .map_err(Error::Call)?;
        if r.outcome == "data" {
            let c = r.chunk.as_ref().ok_or(Error::Invalid("missing chunk"))?;
            require(c.offset == offset && c.total >= offset, "chunk range")?;
            let n = c.data.len() as i64;
            require(
                n <= max_bytes
                    && n <= c.total - offset
                    && c.eof == (n == c.total - offset)
                    && (n > 0 || c.eof),
                "chunk bounds",
            )?;
            self.receipt(&c.receipt, wanted, &self.required)?;
            self.pin(&c.receipt.logical_owner)?;
        } else {
            require(
                matches!(
                    r.outcome.as_str(),
                    "not_ready"
                        | "unavailable"
                        | "unsupported"
                        | "unknown"
                        | "forbidden"
                        | "invalid"
                ) && r.chunk.is_none(),
                "result outcome",
            )?;
        }
        Ok(r)
    }
    pub fn copy_result<W: Write>(
        &self,
        wanted: &wire::RequestIdentity,
        destination: &mut W,
    ) -> Result<u64, CopyError<T::Error>>
    where
        T: abstraction_facade_service::ScopedTransport,
    {
        let transport = self.transport.call_scope().map_err(|cause| CopyError {
            confirmed: 0,
            cause: Error::Transport(cause),
        })?;
        Jobs {
            transport,
            owner: self.owner.clone(),
            required: self.required.clone(),
        }
        .copy_scoped(wanted, destination)
    }
    fn copy_scoped<W: Write>(
        &self,
        wanted: &wire::RequestIdentity,
        destination: &mut W,
    ) -> Result<u64, CopyError<T::Error>> {
        let mut confirmed = 0u64;
        let mut saved = None;
        let result = (|| loop {
            let r = self.read_result(wanted, confirmed as i64, 65536)?;
            if r.outcome != "data" {
                return Err(Error::Outcome(r.outcome));
            }
            let c = r.chunk.unwrap();
            let signature = (c.receipt.operation_id, c.total);
            if let Some(old) = &saved {
                require(*old == signature, "result changed")?;
            } else {
                saved = Some(signature);
            }
            let mut rest = c.data.as_slice();
            while !rest.is_empty() {
                match destination.write(rest) {
                    Ok(0) => return Err(Error::Writer(std::io::ErrorKind::WriteZero.into())),
                    Ok(n) if n <= rest.len() => {
                        confirmed += n as u64;
                        rest = &rest[n..];
                    }
                    Ok(_) => return Err(Error::Invalid("writer count")),
                    Err(e) => return Err(Error::Writer(e)),
                }
            }
            if c.eof {
                return Ok(confirmed);
            }
        })();
        result.map_err(|cause| CopyError { confirmed, cause })
    }
}
/// Inventory is a separate contract; older receipts need not carry new read guarantees.
pub struct Inventory<T>(pub(crate) Jobs<T>);
impl<T: wire::FrameTransport + Clone> Inventory<T> {
    pub fn new(transport: T, owner: String) -> Result<Self, Error<T::Error>> {
        Ok(Self(Jobs::new(transport, owner, vec![])?))
    }
    pub fn list(&self, cursor: &str, limit: i64) -> Result<wire::InventoryPage, Error<T::Error>> {
        require(
            cursor.len() <= 128 && (1..=64).contains(&limit),
            "inventory request",
        )?;
        let p = wire::JobInventoryClient::new(self.0.transport.clone())
            .ListWork(cursor.into(), limit)
            .map_err(Error::Call)?;
        if p.outcome != "page" {
            require(
                matches!(
                    p.outcome.as_str(),
                    "gap" | "forbidden" | "invalid" | "unavailable"
                ) && p.snapshots.is_empty()
                    && p.next.is_empty()
                    && !p.complete,
                "inventory outcome",
            )?;
            return Ok(p);
        }
        require(
            p.snapshots.len() <= limit as usize
                && p.next.len() <= 128
                && if p.complete {
                    p.next.is_empty()
                } else {
                    !p.next.is_empty() && p.next != cursor
                },
            "inventory continuation",
        )?;
        let mut seen = std::collections::HashSet::new();
        let mut owner = None;
        for s in &p.snapshots {
            self.0.snapshot(s, &s.receipt.identity, &[])?;
            require(
                seen.insert(&s.receipt.operation_id)
                    && owner.map_or(true, |o| o == &s.receipt.logical_owner),
                "inventory receipts",
            )?;
            owner = Some(&s.receipt.logical_owner);
        }
        if let Some(o) = owner {
            self.0.pin(o)?;
        }
        Ok(p)
    }
}

impl<C: super::Connector> Jobs<super::Binding<C>> {
    /// Restore caller-retained binding using its trusted connector, without discovery.
    pub fn restore(
        connector: C,
        endpoint: &str,
        owner: String,
        required: Vec<String>,
        deadline: Option<std::time::Instant>,
        cancellation: Option<C::Cancellation>,
    ) -> Result<Self, Error<super::TransportError<C>>> {
        require(!owner.is_empty(), "retained owner")?;
        let binding =
            super::Binding::restore(connector, endpoint, deadline, cancellation, 2 * 1024 * 1024)
                .map_err(|e| Error::Call(wire::CallError::Transport(e)))?;
        Self::new(binding, owner, required)
    }
    pub fn endpoint(&self) -> &str {
        self.transport.endpoint()
    }
    /// Fresh waiting policy through the shared binding; logical ownership stays pinned.
    pub fn with_waiting(
        &self,
        deadline: Option<std::time::Instant>,
        cancellation: Option<C::Cancellation>,
    ) -> Result<Self, Error<super::TransportError<C>>> {
        let binding = self
            .transport
            .with_waiting(deadline, cancellation)
            .map_err(|e| Error::Call(wire::CallError::Transport(e)))?;
        let mut result = Self::new(binding, self.owner(), self.required.clone())?;
        result.owner = self.owner.clone();
        Ok(result)
    }
}

#[cfg(test)]
mod tests;

impl<E: std::fmt::Debug> std::fmt::Display for Error<E> {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{self:?}")
    }
}
impl<E: std::fmt::Debug> std::error::Error for Error<E> {}
impl<E: std::fmt::Debug> std::fmt::Display for CopyError<E> {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(
            f,
            "after {} confirmed bytes: {}",
            self.confirmed, self.cause
        )
    }
}
impl<E: std::fmt::Debug> std::error::Error for CopyError<E> {}
impl<C: super::Connector> Inventory<super::Binding<C>> {
    pub fn endpoint(&self) -> &str {
        self.0.endpoint()
    }
    pub fn owner(&self) -> String {
        self.0.owner()
    }
    pub fn with_waiting(
        &self,
        deadline: Option<std::time::Instant>,
        cancellation: Option<C::Cancellation>,
    ) -> Result<Self, Error<super::TransportError<C>>> {
        Ok(Self(self.0.with_waiting(deadline, cancellation)?))
    }
}
